Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+        this.camera = camera;
+
+        this.clearCommand = new ClearCommand({
+            depth : 1.0,
+            renderState : this.renderState,
+            framebuffer : undefined, // Set later
+            owner : this
+        });
+    }
+
+    function destroyTextures(shadowMap) {
+        shadowMap._depthStencilTexture = shadowMap._depthStencilTexture && !shadowMap._depthStencilTexture.isDestroyed() && shadowMap._depthStencilTexture.destroy();
+    }
+
+    function destroyFramebuffers(shadowMap) {
+        shadowMap.framebuffer = shadowMap.framebuffer && !shadowMap.framebuffer.isDestroyed() && shadowMap.framebuffer.destroy();
+    }
+
+    function createTextures(shadowMap, context) {
+        // TODO : Use nearest filtering for testing, not default linear.
+        shadowMap._depthStencilTexture = new Texture({
+            context : context,
+            width : shadowMap._size,
+            height : shadowMap._size,
+            pixelFormat : PixelFormat.DEPTH_STENCIL,
+            pixelDatatype : PixelDatatype.UNSIGNED_INT_24_8
+        });
+    }
+
+    function createFramebuffers(shadowMap, context) {
+        destroyTextures(shadowMap);
+        destroyFramebuffers(shadowMap);
+
+        createTextures(shadowMap, context);
+
+        var framebuffer = new Framebuffer({
+            context : context,
+            depthStencilTexture : shadowMap._depthStencilTexture,
+            destroyAttachments : false
+        });
+
+        shadowMap._framebuffer = framebuffer;
+        shadowMap.passState.framebuffer = framebuffer;
+        shadowMap.clearCommand.framebuffer = framebuffer;
+    }
+
+    function updateFramebuffers(shadowMap, context) {
+        var depthStencilTexture = shadowMap._depthStencilTexture;
+        var textureChanged = !defined(depthStencilTexture) || (depthStencilTexture.width !== shadowMap._size);
+        if (!defined(shadowMap.framebuffer) || textureChanged) {
+            createFramebuffers(shadowMap, context);
+        }
+    }
+
+    ShadowMap.createShadowCastProgram = function(shaderProgram) {
+        // TODO : unused right now
+        // TODO : vertex shader won't be optimized
+        // TODO : handle mismatched varyings
+        var vs = shaderProgram.vertexShaderText;
+        var fs =
+            'void main()\n' +
+            '{\n' +
+            '    gl_FragColor = vec4(0.0);\n' +
+            '}\n';
+
+        return ShaderProgram.fromCache({
+            vertexShaderSource : vs,
+            fragmentShaderSource : fs,
+            attributeLocations : shaderProgram._attributeLocations
+        });
+    };
+
+    ShadowMap.createReceiveShadowsVertexShader = function(vs) {
+        vs = ShaderSource.replaceMain(vs, 'czm_shadow_main');
+        vs +=
+            'varying vec3 czm_shadowMapCoordinate; \n' +
+            'void main() \n' +
+            '{ \n' +
+            '    czm_shadow_main(); \n' +
+            '    czm_shadowMapCoordinate = (czm_shadowMapMatrix * gl_Position).xyz; \n' +
```

Comment:
TODO: will need to adapt this for GPU RTE, e.g., when there is low and high position attributes.